jupyterlab/loader.js

```diff
@@ -0,0 +1,227 @@
+
+
+/**
+ * Define a module that can be synchronously required.
+ *
+ * @param name - The version-mangled fully qualified name of the module.
+ *   For example, "foo@1.0.1/lib/bar/baz.js".
+ *
+ * @param callback - The callback function for invoking the module.
+ */
+function defineModule(name, callback) {
+  if (!(name in registered)) {
+    registered[name] = callback;
+  }
+}
+
+
+/**
+ * Synchronously require a module that has already been loaded.
+ *
+ * @param name - The semver-mangled fully qualified name of the module.
+ *   For example, "foo@^1.1.0/lib/bar/baz.js".
+ *
+ * @returns The exports of the requested module, if registered.  The module
+ *   selected is the registered module that maximally satisfies the semver
+ *   range of the request.
+ */
+function requireModule(name) {
+  // Check if module is in cache.
+  var moduleId = findModuleId(name);
+  if(installedModules[moduleId]) {
+    return installedModules[moduleId].exports;
+  }
+
+  // Create a new module (and put it into the cache).
+  var module = installedModules[moduleId] = {
+    exports: {},
+    id: moduleId,
+    loaded: false
+  };
+
+  // Execute the module function.
+  registered[moduleId].call(module.exports, module, module.exports, requireModule);
+
+  // Flag the module as loaded.
+  module.loaded = true;
+
+  // Return the exports of the module.
+  return module.exports;
+}
+
+
+/**
+ * Ensure a bundle is loaded on a page.
+ *
+ * @param path - The public path of the bundle (e.g. "lab/jupyter.bundle.js").
+ *
+ * @param callback - The callback invoked when the bundle has loaded.
+ */
+function ensureBundle(path, callback) {
+  // "0" is the signal for "already loaded"
+  if (installedBundles[path] === 0) {
+    return callback.call(null, requireModule);
+  }
+
+  // An array means "currently loading".
+  if (Array.isArray(installedBundles[path])) {
+    installedBundles[path].push(callback);
+    return;
+  }
+
+  // Start bundle loading.
+  installedBundles[path] = [callback];
+  var head = document.getElementsByTagName('head')[0];
+  var script = document.createElement('script');
+  script.type = 'text/javascript';
+  script.charset = 'utf-8';
+  script.async = true;
+  script.onload = function() {
+    var callbacks = installedBundles[path];
+    while(callbacks.length)
+      callbacks.shift().call(null, requireModule);
+    installedBundles[path] = 0;
+  }
+  head.appendChild(script);
+  script.src = path;
+}
+
+
+/**
+ * Ensure a list of bundles are loaded on the page.
+ *
+ * @param paths - The public paths of the bundles. (e.g. "lab/foo.bundle.js").
+ *
+ * @returns A promise that resolves with the require function when all the
+ *   bundles have loaded.
+ */
+function requireBundles(paths) {
+  var promises = paths.map(function(path) {
+    return new Promise(function (resolve, reject) {
+      ensureBundle(path, resolve);
+    });
+  });
+  return Promise.all(promises).then(function () {
+    return requireModule;
+  });
+}
```

Comment:
Since the ensureBundle function doesn't handle errors, we can't ever have this promise be chained with a `.catch` - instead it just hangs. I think it would be better to figure out a way to get an error.